internal/controller/database_controller.go

```diff
@@ -110,63 +113,52 @@ func (r *DatabaseReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 
 	// Check if the Database is marked to be deleted, which is
 	// indicated by the deletion timestamp being set.
-	isDatabaseMarkedToBeDeleted := dbcr.GetDeletionTimestamp() != nil
-	if isDatabaseMarkedToBeDeleted {
-		dbcr.Status.Phase = dbPhaseDelete
-		// Run finalization logic for database. If the
-		// finalization logic fails, don't remove the finalizer so
-		// that we can retry during the next reconciliation.
-		if sliceContainsSubString(dbcr.ObjectMeta.Finalizers, "dbuser.") {
-			err := errors.New("database can't be removed, while there are DbUser referencing it")
-			logrus.Error(err)
-			return r.manageError(ctx, dbcr, err, true)
-		}
-		if containsString(dbcr.ObjectMeta.Finalizers, "db."+dbcr.Name) {
-			err := r.deleteDatabase(ctx, dbcr)
-			if err != nil {
-				logrus.Errorf("DB: namespace=%s, name=%s failed deleting database - %s", dbcr.Namespace, dbcr.Name, err)
-				// when database deletion failed, don't requeue request. to prevent exceeding api limit (ex: against google api)
-				return r.manageError(ctx, dbcr, err, false)
-			}
-			kci.RemoveFinalizer(&dbcr.ObjectMeta, "db."+dbcr.Name)
-			err = r.Update(ctx, dbcr)
-			if err != nil {
-				logrus.Errorf("error resource updating - %s", err)
+	if dbcr.GetDeletionTimestamp() != nil {
+		present = false
+		dbcr.Status.Status = false
+	}
+	// All the k9s objects must be updated only once per reconciliaion
+
+	// Secret
+	// If created -> set the owner annotation
+	// If exists ->
+	//   if not used by other resources -> Update and set usedBy labels
+	var dbSecret *corev1.Secret
+	dbSecret, err = r.getDatabaseSecret(ctx, dbcr)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			if err := r.initialize(ctx, dbcr); err != nil {
 				return r.manageError(ctx, dbcr, err, true)
 			}
-		}
-		// legacy finalizer just remove
-		// we set owner reference for monitoring related resource instead of handling finalizer
-		if containsString(dbcr.ObjectMeta.Finalizers, "monitoring."+dbcr.Name) {
-			kci.RemoveFinalizer(&dbcr.ObjectMeta, "monitoring."+dbcr.Name)
-			err = r.Update(ctx, dbcr)
+			dbSecret, err = r.createSecret(ctx, dbcr)
 			if err != nil {
-				logrus.Errorf("error resource updating - %s", err)
-				return r.manageError(ctx, dbcr, err, true)
+				return r.manageError(ctx, dbcr, err, false)
 			}
+		} else {
+			return r.manageError(ctx, dbcr, err, false)
 		}
-		return reconcileResult, nil
 	}
 
-	databaseSecret, err := r.getDatabaseSecret(ctx, dbcr)
-	if err != nil && !k8serrors.IsNotFound(err) {
-		logrus.Errorf("could not get database secret - %s", err)
+	// Create or Update the secret, if database is removed, labels, should be removed
+	if err := kubehelper.HandleObjectModification(ctx, r.Client, dbSecret, dbcr, dbcr.Spec.Cleanup, present); err != nil {
```

Comment:
all other call this in there related phase/resource function. could we also do this here?